File: haskell/allergies/Allergies.hs
module Allergies (Allergen(..), isAllergicTo, allergies) where

import Data.Bits

data Allergen = Eggs 
              | Peanuts 
              | Shellfish 
              | Strawberries 
              | Tomatoes 
              | Chocolate 
              | Pollen 
              | Cats 
              deriving (Show, Eq, Enum, Bounded)

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo allergy score = testBit score allergyBit
                           where allergyBit = fromEnum allergy

allergies :: Int -> [Allergen]
allergies score = filter scoreAllergicTo allAllergies
                where scoreAllergicTo = flip isAllergicTo score
                      allAllergies    = enumFrom minBound

File: csharp/allergies/Allergies.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

public class Allergies
{
    private enum Allergen { eggs = 1, peanuts = 2, shellfish = 4, strawberries = 8, tomatoes = 16, chocolate = 32, pollen = 64, cats = 128 }
    private List<string> _allergies;
    private int _score;

    public Allergies(int score)
    {
        _score = score;
        _allergies = new List<string>();
        foreach (var value in Enum.GetValues(typeof(Allergen)))
        {
            if ((_score & (int)value) == (int)value)
            {
                _allergies.Add(value.ToString());
            }
        }
    }

    public bool AllergicTo(string allergen)
    {
        return _allergies.Contains(allergen);
    }

    public List<string> List()
    {
        return _allergies;
    }
}

